ledger/src/slot_stats.rs

```diff
@@ -0,0 +1,90 @@
+use {
+    crate::blockstore_meta::SlotMeta, bitflags::bitflags, lru::LruCache, solana_sdk::clock::Slot,
+};
+
+const SLOTS_STATS_CACHE_CAPACITY: usize = 300;
+
+macro_rules! get_mut_entry (
+    ($cache:expr, $key:expr) => (
+        match $cache.get_mut(&$key) {
+            Some(entry) => entry,
+            None => {
+                $cache.put($key, SlotStats::default());
+                $cache.get_mut(&$key).unwrap()
+            }
+        }
+    );
+);
+
+#[derive(Copy, Clone, Debug)]
+pub(crate) enum ShredSource {
+    Turbine,
+    Repaired,
+    Recovered,
+}
+
+bitflags! {
+    #[derive(Default)]
+    struct SlotFlags: u8 {
+        const DEAD   = 0b00000001;
+        const FULL   = 0b00000010;
+        const ROOTED = 0b00000100;
+    }
+}
+
+#[derive(Default)]
+struct SlotStats {
+    flags: SlotFlags,
+    num_repaired: usize,
+    num_recovered: usize,
+}
+
+pub(crate) struct SlotsStats(LruCache<Slot, SlotStats>);
+
+impl Default for SlotsStats {
+    fn default() -> Self {
+        // LruCache::unbounded because capacity is enforced manually.
+        Self(LruCache::unbounded())
+    }
+}
+
+impl SlotsStats {
+    pub(crate) fn add_shred(&mut self, slot: Slot, source: ShredSource) {
```

Comment:
nit: add_shred -> add_shred_source

Reply:
I am expecting with @jbiseda work we are going to add more shred attributes here than just the source. The erasure batch index in particular.

Reply:
makes sense!